src/support/modify.c

```diff
@@ -8,28 +8,58 @@
 
 #include "wt_internal.h"
 
+#define	WT_MODIFY_FOREACH_BEGIN(mod, p, nentries) do {			\
+	const size_t *__p = p;						\
+	const uint8_t *__data =						\
+	    (const uint8_t *)(__p + (size_t)nentries * 3);		\
+	int __i;							\
+	for (__i = 0; __i < nentries; ++__i) {				\
+		mod.data.data = __data;					\
+		memcpy(&mod.data.size, __p++, sizeof(size_t));		\
+		__data += mod.data.size;				\
+		memcpy(&mod.offset, __p++, sizeof(size_t));		\
+		memcpy(&mod.size, __p++, sizeof(size_t));
+
+#define	WT_MODIFY_FOREACH_REVERSE(mod, p, nentries, datasz) do {	\
+	const size_t *__p = p + (size_t)nentries * 3;			\
+	const uint8_t *__data = (const uint8_t *)__p + datasz;		\
+	int __i;							\
+	for (__i = 0; __i < nentries; ++__i) {				\
+		memcpy(&mod.size, --__p, sizeof(size_t));		\
+		memcpy(&mod.offset, --__p, sizeof(size_t));		\
+		memcpy(&mod.data.size, --__p, sizeof(size_t));		\
+		mod.data.data = (__data -= mod.data.size);
+
+#define	WT_MODIFY_FOREACH_END						\
+	}								\
+} while (0)
+
```

Comment:
Why did you decide to copy this information twice? Unless we expect large numbers of modifications in the common case (where we'd have to allocate a `WT_MODIFY` structure), why not allocate an array on the stack as the previous version did?

Reply:
I'm copying it less ;-)

Seriously, the inner loop of this macro turns into 4 instructions:
```
       addq    $24, %r12       #, p
        movq    -24(%r12), %r13 # MEM[base: p_130, offset: 0B], D.42154
        movq    -16(%r12), %rbx # MEM[base: p_130, offset: 8B], D.42154
        movq    -8(%r12), %rcx  # MEM[base: p_195, offset: -8B], size
```

Previously, this was first copied to the stack, then copied out of there when needed.  There was a threshold size where `malloc` would kick in and the performance would change.  This version doesn't make a copy and so never needs to allocate extra space to hold it.